### ISSUE TITLE
script(1)/su(1) child wait

### ISSUE THREAD

**Issue by 8vasu** [karelzak/util-linux#1145]
This follows pull request #1141 

One comment by me says

"I too think that it should be the program's responsibility. Especially upon error. However, I think that the main loop actually does wait for the child in many cases: see "case SIGCHLD" in handle_signal(). Moreover, ul_pty_set_child(pty, (pid_t) -1) is called in ul_pty_wait_for_child(); because of this, a custom wait in main() might be skipped. I am not sure about what happens in login-utils/su-common.c, but we should check to make sure that it does not skip its custom wait. If it does skip the custom wait, then we should consider removing the wait block from "case SIGCHLD" and only set poll timeout there; we should then mandate that each program using lib/pty-session.c wait for the child in main() after ul_pty_proxy_master(). Of course, we can use pty->callbacks.child_wait() to set the custom wait to be called in main(). Please take your time to verify these things :)"

In su-common.c,

		if (ul_pty_proxy_master(su->pty) != 0)
			caught_signal = true;

		/* ul_pty_proxy_master() keeps classic signal handler are out of game */
		caught_signal = ul_pty_get_delivered_signal(su->pty);

the "if" statement is redundant because caught_signal is reassigned immediately after that.

Also, in su-common.c,

	/*
	 * Wait for child
	 */
	if (!caught_signal)
		status = wait_for_child(su);
	else
		status = 1;

the "status = wait_for_child(su);" is wasted when everything goes well and the child exits normally; this is because handle_signal() in lib/pty-session.c already waits for child upon SIGCHLD and sets pty->child to -1.
I suspect that a similar thing happens with the /* final wait */ in script(1).

Please test this hypothesis. I have not found time to test it yet. I can make another pull request after the issue is assessed properly.

**Comment by karelzak** [comment]
> One comment by me says

I know, I had no time to play with it yet. Thanks you added this issue as a reminder.

> In su-common.c,
> 
> ```
> 	if (ul_pty_proxy_master(su->pty) != 0)
> 		caught_signal = true;
>
> 	/* ul_pty_proxy_master() keeps classic signal handler are out of game */
> 	caught_signal = ul_pty_get_delivered_signal(su->pty);
> ```
> 
> the "if" statement is redundant because caught_signal is reassigned immediately after that.

Yes, the `if()` is nonsense. Fixed.

> ```
> /*
>  * Wait for child
>  */
> if (!caught_signal)
> 	status = wait_for_child(su);
> else
> 	status = 1;
> ```
> 
> the "status = wait_for_child(su);" is wasted when everything goes well and the child exits normally; this is because handle_signal() in lib/pty-session.c already waits for child upon SIGCHLD and sets pty->child to -1.

It's expected and wanted to call su-common.c:wait_for_child() more than once :-) The first call from pty-session.c:handle_signal() sets su->child = -1 and su->childstatus according status from waitpid(). The second call from the main() returns only su->childstatus and nothing else if child = -1.

The idea is to keep wait_for_child() usable when su(1) executed with and without --pty to get compatible results in both cases and to keep the final status calculation on one place as I want to minimize --pty specific code in su(1).

> I suspect that a similar thing happens with the /* final wait */ in script(1).

The final wait in script.c:main() means that we still have a child because ctl.child is not -1. The ->child in su and script is 
set to -1 only after sucessful waitpid(), otherwise we have to care about the child (wait or kill).
```
       if (!caught_signal && ctl.child != (pid_t)-1)
                ul_pty_wait_for_child(ctl.pty); /* final wait */
```

Note that in both tools we kill the child if we get a signal (caught_signal is set) in parental process.

> Please test this hypothesis. I have not found time to test it yet. I can make another pull request after the issue is assessed properly.

It all seems correct from my point of view (weel, except the `if()`).

**Comment by 8vasu** [comment]
Thank you :D I got it! This issue is now resolved.
